File: backend/app/models/doubt.py
from sqlalchemy import Column, Integer, String, ForeignKey, Text, DateTime, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.session import Base
import enum

class DoubtStatus(str, enum.Enum):
    PENDING = "pending"
    ANSWERED = "answered"

class Doubt(Base):
    __tablename__ = "doubts"

    id = Column(Integer, primary_key=True, index=True)
    student_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    course_id = Column(Integer, ForeignKey("courses.id"), nullable=False)
    question_text = Column(Text, nullable=False)
    teacher_reply = Column(Text, nullable=True)
    status = Column(Enum(DoubtStatus), default=DoubtStatus.PENDING)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # Relationships
    student = relationship("User", backref="doubts_asked")
    course = relationship("Course", backref="course_doubts")

class DoubtSession(Base):
    __tablename__ = "doubt_sessions"

    id = Column(Integer, primary_key=True, index=True)
    course_id = Column(Integer, ForeignKey("courses.id"), nullable=False)
    teacher_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    session_date = Column(DateTime(timezone=True), nullable=False)
    meeting_link = Column(String, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    course = relationship("Course", backref="doubt_sessions")
    teacher = relationship("User", backref="sessions_hosted")

File: frontend/src/services/doubt.ts
import api from "@/lib/api";

export interface Doubt {
    id: number;
    student_id: number;
    course_id: number;
    question_text: string;
    teacher_reply?: string;
    status: "pending" | "answered";
    created_at: string;
    student_name?: string;
}

export interface Session {
    id: number;
    course_id: number;
    teacher_id: number;
    session_date: string;
    meeting_link: string;
}

export const doubtService = {
    createDoubt: async (course_id: number, question_text: string) => {
        const response = await api.post<Doubt>("/doubts/", {
            course_id,
            question_text,
        });
        return response.data;
    },

    getDoubts: async (course_id: number) => {
        const response = await api.get<Doubt[]>(`/doubts/course/${course_id}`);
        return response.data;
    },

    replyDoubt: async (doubt_id: number, teacher_reply: string) => {
        const response = await api.put<Doubt>(`/doubts/${doubt_id}/reply`, {
            teacher_reply,
            status: "answered", // Backend defaults, but explicit is fine
        });
        return response.data;
    },

    // LIVE SESSIONS
    scheduleSession: async (course_id: number, session_date: string, meeting_link: string) => {
        const response = await api.post<Session>("/doubts/session", {
            course_id,
            session_date,
            meeting_link
        });
        return response.data;
    },

    getSessions: async (course_id: number) => {
        const response = await api.get<Session[]>(`/doubts/session/course/${course_id}`);
        return response.data;
    }
};
